### ISSUE TITLE
Notifications emails are not sent due to invalid From

### ISSUE THREAD

**Issue by strk** [gogits/gogs#3615]
Noticing I was not receiving email notifications I checked the logs and found:

```
gomail: could not send email 1: gomail: invalid address "\"bjornharrtell\" <>": mail: invalid string
```

As the user `bjornharrtell` has a valid email registered (checked via web UI and on the database), I went looking at code and found this in modules/log/smtp.go:

```
    mailmsg := []byte("To: " + strings.Join(s.RecipientAddresses, ";") + "\r\nFrom: " + s.Username + "<" + s.Username +
        ">\r\nSubject: " + s.Subject + "\r\n" + content_type + "\r\n\r\n" + fmt.Sprintf(".%s", time.Now().Format("2006-01-02 15:04:05")) + msg)

    return smtp.SendMail(                                                       
        s.Host,                                                                 
        auth,                                                                   
        s.Username,                                                             
        s.RecipientAddresses,                                                   
        mailmsg,                                                                
    )             
```

Note how `s.Username` is used for the `From` field both in the mail header and in the `smtp.SendMail` 3rd parameter.

I suspect `smtp.SendMail` just refuses such `From` address (lacking an `@` character) and fails to send anything.

I believe things did work as of 0.9.68, while are failing as of 0.9.99

**Comment by strk** [comment]
I guess the `From` value should be set to the `FROM` parameter under `mailer` config section ?

**Comment by strk** [comment]
Ok I changed analisys, now I think the problem is with `composeIssueMessage` in `models/mail.go`:
```
func composeIssueMessage(issue *Issue, doer *User, tplName base.TplName, tos []string, info string) *mailer.Message {
    subject := issue.MailSubject()                                              
    body := string(markdown.RenderSpecialLink([]byte(issue.Content), issue.Repo.HTMLURL(), issue.Repo.ComposeMetas()))
    data := composeTplData(subject, body, issue.HTMLURL())                      
    data["Doer"] = doer                                                         
    content, err := mailRender.HTMLString(string(tplName), data)                
    if err != nil {                                                             
        log.Error(3, "HTMLString (%s): %v", tplName, err)                       
    }                                                                           
    msg := mailer.NewMessageFrom(tos, fmt.Sprintf(`"%s" <%s>`, doer.DisplayName(), setting.MailService.User), subject, content)
    msg.Info = fmt.Sprintf("Subject: %s, %s", subject, info)                    
    return msg                                                                  
}       
```

The `NewMessageFrom` method second parameter sets the `From` line, and the code is using `setting.MailService.User` rather than `setting.MailService.From`. The former should be used for SMTP authentication, when needed, while the latter is the actual `From`. In my deploy there's no `User` as no authentication is needed.

**Comment by strk** [comment]
Confirmed, setting a `USER` variable in the configuration is an effective workaround.
I will file a pull request with a fix to the bug.

**Comment by DblK** [comment]
From my point of view, the [SmtpWriter](https://github.com/gogits/gogs/blob/master/modules/log/smtp.go#L20), need to have also the Username instead of using the field Username as UserEmail.

After adding this information, the [from String](https://github.com/gogits/gogs/blob/master/modules/log/smtp.go#L67) needs to be changed according to the new field in the structure.

That way, according to the source, it will be `From: Example <example@gmail.com>` which is better than `From: example@gmail.com <example@gmail.com>`.

**Comment by strk** [comment]
You meant `need to have also the UserEmail` I guess ? If I read the code correctly `SmtpWriter` has a `From` configured only once, so it should be `settings.MailService.From`, producing `From: User Name <system@from.dns>`. Or are you saing that the `s.Username` value is really an email ?

Anyway, this ticket was for issue notifications, which are fixed by PR #3616, if you can see a bug due to `SmtpWriter` please file a separate ticket, with instructions to reproduce. Thanks

**Comment by DblK** [comment]
According to the [comments](https://github.com/gogits/gogs/blob/master/modules/log/smtp.go#L37), `Username` is an UserEmail and if it's not the email could not be sent.
I will fill another issue form that specific.

**Comment by Unknwon** [comment]
Merge this thread to https://github.com/gogits/gogs/issues/3856.
